### PR TITLE
Separate textstat

### PR DESCRIPTION
**Pull request by kbenoit**
This is part of a multi-step process to modularise the package. 

- Create a branch called `modularise` from the master. This will be the branch that will work with the modularised packages, and from which we successively split off from the main branch.

- Implement the modularisation.
    - [**quanteda.textmodels**](https://github.com/quanteda/quanteda.textmodels): This is already split and already on CRAN. It might have its textplot function(s) moved, but is more or less fine. 
    - [**quanteda.textstats**](https://github.com/quanteda/quanteda.textstats): This is the package containing all textstat functions. We need to split it and then do a full set of revdep checks on the pared down **quanteda**, to see how much we have broken, and then fix these
    - **quanteda.textplots**: Once the textstats split is working, we can then perform the next split, moving the textplot functions to this new package. This will be slightly different in that it will Import all three other packages (**quanteda**, **quanteda.textmodels**, and **quanteda.textstats**) since it operates on objects from all three. But we should think carefully about this since it could be that we would like these functions to work on non-quanteda objects too.

- Submit these to CRAN as part of a careful update strategy. The big problem will be finding and fixing the use of `quanteda::movedfunction()` (e.g. `quanteda::textstat_collocations()` for instance, since it will have moved to another namespace.

- Fix the pkgdown documentation and vignettes. It would be nice to have all of these in one place, as now, but also make it clear that they are divided among different packages. See https://github.com/r-lib/pkgdown/issues/1251.

- Add more packages to the quanteda family, such as **quanteda.sentiment**, **quanteda.newsmap**, etc.

**Note:** @koheiw got this branch by merging the `modularise` branch from [**quanteda.core**](https://github.com/koheiw/quanteda.core/pull/5). Because it was a copy from **quanteda** I was able to merge it and then tidy it up. I restored any functions not part of the textstat separation - we should only separate one thing at a time. This included the textplot functions but also some others that had vanished for reasons unknown - `bootstrap_dfm()`, and other things we had recently added like `dictionary_edit()`, `char_select()`, etc.

### CODE REVIEW

File: R/char_select.R [hunk 1]
```diff
@@ -44,7 +44,7 @@ char_select.character <- function(x, pattern, selection = c("keep", "remove"),
     
     if (is.dictionary(pattern))
         pattern <- unique(unlist(as.list(pattern), use.names = FALSE))
-    if (is.collocations(pattern))
+    if ("collocations" %in% class(pattern))
```
Comment (koheiw) [inline R/char_select.R:47]: I think it is better to have object checking functions in the core package at least as internals.

File: R/summary.R [hunk 2]
```diff
@@ -25,7 +25,7 @@ summary.corpus <- function(object, n = 100, tolower = FALSE, showmeta = TRUE, ..
     ndoc_all <- ndoc(object)
     object <- head(object, n)
     ndoc_show <- ndoc(object)
-    result <- textstat_summary(object, cache = FALSE, tolower = tolower, ...)
+    result <- summarize(as.corpus(object), cache = FALSE, tolower = tolower, ...)
```
Comment (koheiw) [inline R/summary.R:28]: Let's make summary to be like 
```
> summary(attr(data_corpus_inaugural, "docvars"))
   docname_                     docid_       segid_       Year     
 Length:58          1789-Washington: 1   Min.   :1   Min.   :1789  
 Class :character   1793-Washington: 1   1st Qu.:1   1st Qu.:1846  
 Mode  :character   1797-Adams     : 1   Median :1   Median :1903  
                    1801-Jefferson : 1   Mean   :1   Mean   :1903  
                    1805-Jefferson : 1   3rd Qu.:1   3rd Qu.:1960  
                    1809-Madison   : 1   Max.   :1   Max.   :2017  
                    (Other)        :52                             
  President          FirstName                           Party   
 Length:58          Length:58          Democratic           :21  
 Class :character   Class :character   Democratic-Republican: 7  
 Mode  :character   Mode  :character   Federalist           : 1  
                                       none                 : 2  
                                       Republican           :24  
                                       Whig                 : 3  
```

File: R/summary.R [hunk 3]
```diff
@@ -59,4 +59,94 @@ print.summary.corpus <- function(x, ...) {
...
+        result$chars <- nchar(x)
+        result$sents <- ntoken(tokens(x, what = "sentence"))
+    }
+
+    if (cache)
+        set_cache(x, "summary", result, ...)
+    return(result)
+}
+
+get_cache <- function(x, field, ...) {
```
Comment (koheiw) [inline R/summary.R:114]: Better to move them to `cache.R`

File: tests/testthat/test-tokens_compound.R [hunk 4]
```diff
@@ -129,42 +129,6 @@ test_that("tokens_compound works as expected with nested and overlapping tokens"
     )
 })
 
-test_that("tokens_compound works as expected with collocations", {
```
Comment (koheiw) [inline tests/testthat/test-tokens_compound.R:132]: Better not delete. We can test `tokens_compound()` with a save collocation object.

File: tests/testthat/test-tokens_select.R [hunk 5]
```diff
@@ -218,43 +218,43 @@ test_that("tokens_select works as expected for bigrams selected on char, list of
...
-    )
-    expect_equal(
-        as.list(tokens_select(toks_bi, coll_tri)),
-        list(d1 = character(), d2 = character())
-    )
-    expect_silent(
-        tokens_select(toks_bi, coll_tri)
-    )
-})
+# test_that("tokens_select works correctly with collocations objects", {
```
Comment (koheiw) [inline tests/testthat/test-tokens_select.R:221]: We should test using saved collocation objects.

**Review by koheiw — CHANGES_REQUESTED** [review]
Looks good overall. Only two major issues as commented on the code:

- We should keep the tests of `tokens_select()` and `tokens_compound()` with collocations using saved objects.
- We should really simplify the output of `summary().
-

File: R/char_select.R (hunk 1 above)
Comment (kbenoit) [reply R/char_select.R]: Agreed. Added back in c66f617.

File: R/summary.R (hunk 2 above)
Comment (kbenoit) [reply R/summary.R:28]: The key for the separation is how to wean it from `textstat_summary()`. How we design the summary is better discussed and resolved via #1959.

File: R/summary.R (hunk 3 above)
Comment (kbenoit) [reply R/summary.R:114]: Done.

File: tests/testthat/test-tokens_compound.R (hunk 4 above)
Comment (kbenoit) [reply tests/testthat/test-tokens_compound.R:132]: Agreed. Added back in 9e9aa94. I put them into a separate test file called `test-collocations.R` to make this easily distinct. (This had been in **quanteda.textstats** but I moved it out.)

File: tests/testthat/test-tokens_select.R (hunk 5 above)
Comment (kbenoit) [reply tests/testthat/test-tokens_select.R:221]: Added back in 9e9aa94.

File: R/summary.R (hunk 2 above)
Comment (koheiw) [reply R/summary.R:28]: Duplication is not separation: there are `quanteda::summarize()` and `quanteda.textstats::summarize()`. I am fine if this is only transitional.

**Review by koheiw — APPROVED** [review]

File: R/summary.R (hunk 2 above)
Comment (kbenoit) [reply R/summary.R:28]: Agreed. We will eliminate the duplication by redesigning `summary.corpus()`.

**Pull request merged**